frontend/src/metabase/query_builder/components/dataref/DataReference.jsx

```diff
@@ -33,96 +34,75 @@ const TITLE_ICONS = {
   model: "model",
 };
 
-export default class DataReference extends Component {
-  constructor(props, context) {
-    super(props, context);
-
-    this.state = {
-      stack: this.initialStack(),
-      tables: {},
-      fields: {},
-    };
-  }
+const DataReferencePropTypes = {
+  query: PropTypes.object.isRequired,
+  dataReferenceStack: PropTypes.array.isRequired,
+  setDataReferenceStack: PropTypes.func.isRequired,
+  onClose: PropTypes.func.isRequired,
+  runQuestionQuery: PropTypes.func.isRequired,
+  setDatasetQuery: PropTypes.func.isRequired,
+};
 
-  componentDidUpdate(prevProps) {
-    const { id: dbId } = this.props.query.database() || {};
-    const { id: prevDbId } = prevProps.query.database() || {};
-    if (dbId !== prevDbId) {
-      this.setState({ stack: this.initialStack() });
+const DataReference = ({
+  query,
+  setDataReferenceStack: setStack,
+  dataReferenceStack: stack,
+  onClose,
+  ...props
+}) => {
+  const previousDbId = usePrevious(query?.database()?.id);
+  useEffect(() => {
+    if (query?.database()?.id === previousDbId) {
+      return;
     }
-  }
-
-  initialStack() {
-    const { query } = this.props;
-
     const stack = [];
-    const database = query && query.database();
-    if (database) {
-      stack.push({ type: "database", item: database });
+    if (query?.database()) {
+      stack.push({ type: "database", item: query.database() });
     }
-    const table = query && query.table();
-    if (table) {
-      stack.push({ type: "table", item: table });
+    if (query?.table()) {
+      stack.push({ type: "table", item: query.table() });
     }
-
-    return stack;
+    setStack(stack);
+  });
```

Comment:
Could we move all this logic to redux actions instead? 
- In the NativeQueryEditor we could dispatch an action when `setDatabaseId` is called to initialize the data reference stack for a given query and remove this effect
- `back` and `show` also seem like a regular redux actions changing the state.

Reply:
I think in an ideal world components would get data via selectors and render UI with almost no data manipulation logic (e.g. not UI logic), and callbacks would be bound to redux actions directly. It doesn't make sense to apply this pattern strictly, but it seems to me that it would be a good fit here. Also the existing code isn't the best example currently.